engine/pkg/client/resource_manager_client.go

```diff
@@ -37,21 +39,69 @@ func NewResourceManagerClient(cli enginepb.ResourceManagerClient) ResourceManage
 	return &resourceManagerClient{cli: cli}
 }
 
-func (c *resourceManagerClient) CreateResource(ctx context.Context, request *enginepb.CreateResourceRequest) error {
+func (c *resourceManagerClient) CreateResource(
+	ctx context.Context, request *enginepb.CreateResourceRequest,
+) error {
 	call := internal.NewCall(c.cli.CreateResource, request)
 	_, err := call.Do(ctx)
-	// TODO specialized retry strategy.
-	return err
+	if err == nil {
+		return nil
+	}
+
+	code, ok := rpcerror.GRPCStatusCode(err)
+	if !ok {
+		return err
+	}
+	if code != codes.AlreadyExists {
+		return err
+	}
+
+	originalErr := err
+
+	// If the returned value is that the resource already exists,
+	// we need to do a query and check whether it's a duplicate request.
+	// This is to guarantee idempotency if the network is not stable.
+	queryResp, err := c.QueryResource(ctx, &enginepb.QueryResourceRequest{
+		ResourceKey: &enginepb.ResourceKey{
+			JobId:      request.JobId,
+			ResourceId: request.ResourceId,
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	// Check whether the previous request has actually succeeded.
+	if queryResp.CreatorWorkerId == request.CreatorWorkerId {
+		return nil
+	}
+	return originalErr
```

Comment:
Does this happen when two workers create the same resource?

Reply:
No. It can only happen if the same worker has tried to create the same resource twice.